### PR TITLE
CLAN ARENA: improvements

### PR DESCRIPTION
**Pull request by dusty-qw**

### CODE REVIEW

**Comment by tcsabina** [comment]
Hey dusty,

Why did you close the PR?


Regards,
Toma

**Comment by dusty-qw** [comment]
Hi Toma, 

The upstream PR was an accident :) I meant to merge into my own fork until the work is complete.

**Comment by tcsabina** [comment]
Hi Dusty,
I see the PR...
Give me a couple of days to find some time to give some valuable feedbacks. It looks okay at first glance.
There is only 1 thing I saw:
Many places we call the `isCA()` function, even like 2-3 lines below. Should we consider optimizing this a bit?
A function call is always extra effort. If there is a way to avoid calling the same function relatively close to each other, I think we should do it.
I am not sure what exactly happens, considering compiler optimization, to be honest...

I will give some code samples to be clear about what I mean...


Keep you posted,
Toma

File: src/client.c [hunk 1]
```diff
@@ -4214,6 +4242,10 @@ void PlayerPostThink()
 				self->s.v.ammo_rockets = 100 + (int)(velocity_vert_abs) % 10000 / 100;
 				self->s.v.ammo_cells = 100 + (int)(velocity_vert_abs) % 100;
 			}
+			else if (cvar("k_clan_arena"))
```
Comment (tcsabina) [inline src/client.c:4245]: how come not `isCA()`?

File: src/match.c [hunk 2]
```diff
@@ -439,7 +447,7 @@ void EndMatch(float skip_log)
 	}
 
 	// allow ready/break in bloodfest without map reloading.
-	if (k_bloodfest)
+	if (k_bloodfest || cvar("k_clan_arena"))
```
Comment (tcsabina) [inline src/match.c:450]: Why not `isCA()` call?

File: src/match.c [hunk 3]
```diff
@@ -1387,7 +1399,13 @@ void PrintCountdown(int seconds)
 	}
 	else if (isCA())
 	{
-		mode = redtext("CA");
+		if (cvar("k_clan_arena") == 2)
+		{
+			mode = redtext("Wipeout");	
+		}
+		else {
```
Comment (tcsabina) [inline src/match.c:1406]: small typo. Correctly:
```
else
{
```

File: src/match.c [hunk 4]
```diff
@@ -1973,6 +1991,32 @@ char* CompilateDemoName()
...
+		if (cvar("k_clan_arena") == 1)
+		{
+			strlcat(demoname, "ca", sizeof(demoname));
+		}
+		else
+		{
+			strlcat(demoname, "wipeout", sizeof(demoname));
+		}
+
+		char teams[MAX_CLIENTS][MAX_TEAM_NAME];
```
Comment (tcsabina) [inline src/match.c:2005]: Could you move this to the start of the block, please? I hate to see variable declarations in the middle :).
Is this even compiling? This is not ISO-C, isn't it?

File: src/match.c [hunk 5]
```diff
@@ -1973,6 +1991,32 @@ char* CompilateDemoName()
...
+		getteams(teams);
+		
+		for (vs = "_", i = 0; i < MAX_CLIENTS; i++)
+        {
```
Comment (tcsabina) [inline src/match.c:2009]: small error in the indentation

File: src/match.c [hunk 6]
```diff
@@ -2748,6 +2792,13 @@ void PlayerBreak()
 		return;
 	}
 
+	if ( isCA() && match_in_progress == 2 && !self->ca_ready)
```
Comment (tcsabina) [inline src/match.c:2795]: Correctly:
```
if (isCA() && (match_in_progress == 2) && !self->ca_ready)
```

File: src/commands.c [hunk 7]
```diff
@@ -6362,14 +6372,25 @@ void lastscore_add()
...
+		{
+			lst = lsCTF;
+		}
 
 		e1 = cvar_string("_k_team1");
-		s1 = get_scores1();
+		s1 = isCA() ? CA_get_score_1() : get_scores1();
 		e2 = cvar_string("_k_team2");
-		s2 = get_scores2();
+		s2 = isCA() ? CA_get_score_2() : get_scores2();
```
Comment (tcsabina) [inline src/commands.c:6394]: Here we also have 3 consecutive calls of `isCA()`. Should we not optimize this a bit:
```
	else if ((isTeam() || isCTF() || isCA()) && k_showscores)
	{
		qbool isCa = isCA();
		if (isCa)
		{
			lst = cvar("k_clan_arena") == 2 ? lsWO : lsCA;
		}
		else if (isTeam())
		{
			lst = lsTeam;
		}
		else
		{
			lst = lsCTF;
		}

		e1 = cvar_string("_k_team1");
		s1 = isCa ? CA_get_score_1() : get_scores1();
		e2 = cvar_string("_k_team2");
		s2 = isCa ? CA_get_score_2() : get_scores2();

		// players from first team
		for (t1[0] = from = 0, p = world; (p = find_plrghst(p, &from));)
```


or at least:
```
		e1 = cvar_string("_k_team1");
		e2 = cvar_string("_k_team2");
		if (isCA())
		{
				s1 = CA_get_score_1();
				s2 = CA_get_score_2();
		}
		else
		{
				s1 = get_scores1();
				s2 = get_scores2();
		}
```

File: src/client.c [hunk 8]
```diff
@@ -971,12 +971,17 @@ void ClientKill()
...
+	{
+		G_sprint (self, PRINT_HIGH, "Can't suicide at this time\n");
+
+		return;
+	}
+
+	if (isCA() && match_in_progress && ra_match_fight == 2 && !ca_round_pause)
+	{
+		self->round_deaths = 99;	// No respawning after suicide in wipeout mode
+	}
```
[the remark this answers is not in the document]
Comment (dusty-qw) [reply src/client.c:987]: changed.

File: src/client.c [hunk 9]
```diff
@@ -1398,37 +1407,50 @@ qbool CanConnect()
...
-			if (isTeam() || isCTF())
+			// In CA, if current team doesn't match old team then just don't restore stats/gamestate
+			// Otherwise restore frags and set ca_ready
+			else if (isCA() && strneq(getteam(self), getteam(p)))
 			{
-				self->k_teamnum = p->k_teamnum; // we alredy have team in localinfo
-				G_bprint(2, "%s \220%s\221 %s %d %s%s\n", self->netname, getteam(self),
-							redtext("rejoins the game with"), (int)self->s.v.frags,
-							redtext("frag"), redtext(count_s(self->s.v.frags)));
+				// if player's team isn't what it was before, then he will be a "dead" player until the match is over
```
[the remark this answers is not in the document]
Comment (dusty-qw) [reply src/client.c:1425]: changed.

File: src/client.c (hunk 1 above)
Comment (dusty-qw) [reply src/client.c]: changed.

File: src/match.c (hunk 2 above)
Comment (dusty-qw) [reply src/match.c]: changed.

File: src/match.c (hunk 4 above)
Comment (dusty-qw) [reply src/match.c]: Fixed it. I'm a total C newb so I wouldn't know :) It was compiling though!

File: src/match.c (hunk 5 above)
Comment (dusty-qw) [reply src/match.c]: fixed.

File: src/match.c (hunk 6 above)
Comment (dusty-qw) [reply src/match.c]: changed.

File: src/commands.c (hunk 7 above)
Comment (dusty-qw) [reply src/commands.c:6395]: changed.

File: src/match.c (hunk 4 above)
Comment (tcsabina) [reply src/match.c]: ;)

**Pull request merged**